Extending with custom manual instrumentation (#5789)
open-telemetry/opentelemetry.io: content/en/docs/zero-code/java/agent/api.md

---

> title: Manual instrumentation

Reviewer: ```suggestion
title: Extending with custom manual instrumenation
linkTitle: Extending instrumentation
```

based on @jack-berg's title suggestion

> ---
> title: Extending with custom manual instrumentation using the OpenTelemetry API
> linkTitle: Extending instrumentation
> description: Extending a Java agent with custom manual instrumentation.
> weight: 21
> ---

Reviewer: the title is now longer then the description, here is a proposal

```suggestion
---
title: Extending instrumentations with the API
linkTitle: Extend with the API
description: Use the OpenTelemetry API in combination with the Java agent to extend the automatically generated telemetry with custom instrumentations
weight: 21
---
```

@chalin @jack-berg can you take a look as well?

Reviewer 2: Your proposal LGTM.

Author: I adjusted slightly:

> Use the OpenTelemetry API in combination with the Java agent to extend the automatically generated telemetry with custom spans and metrics

[metrics](/docs/concepts/signals/metrics) for your own code without doing too
many code changes.

## Dependencies

Add a dependency on the `opentelemetry-api` library.

### Maven

```xml
<dependencies>
  <dependency>
    <groupId>io.opentelemetry</groupId>
    <artifactId>opentelemetry-api</artifactId>
    <version>{{% param vers.otel %}}</version>
  </dependency>
</dependencies>
```

### Gradle

```groovy
dependencies {
    implementation('io.opentelemetry:opentelemetry-api:{{% param vers.otel %}}')
}
```

## OpenTelemetry

The Java agent is a special case where `GlobalOpenTelemetry` is set by the
agent. Simply call `GlobalOpenTelemetry.get()` to access the `OpenTelemetry`
instance.

## Span

{{% alert title="Note" color="info" %}}

For the most common use cases, use the `@WithSpan` annotation instead of manual
instrumentation. See [Annotations](../annotations) for more information.

{{% /alert %}}

```java
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.trace.Tracer;

Tracer tracer = GlobalOpenTelemetry.getTracer("application");
```

Use the `Tracer` to create a span as explained in the
[Span](/docs/languages/java/api/#span) section.

> A full example can be found [example repository].

Reviewer 3: ```suggestion
A full example can be found in the [example repository].
```

## Meter

```java
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.metrics.Meter;

Meter meter = GlobalOpenTelemetry.getMeter("application");
```

Use the `Meter` to create a counter, gauge or histogram as explained in the
[Meter](/docs/languages/java/api/#meter) section.

> A full example can be found [example repository].

Reviewer 3: ```suggestion
A full example can be found in the [example repository].
```
